### ISSUE TITLE
[TensorRT-LLM][ERROR] Assertion failed: Size per head is not a multiple of X (unfusedAttentionKernels.cu:1710)

### ISSUE THREAD

**Issue by jpf888** [NVIDIA/TensorRT-LLM#855]
When I quantized int8kv cache on the open source model of openllama 3B, in the last step of run.py, 
`[TensorRT-LLM][ERROR] Assertion failed: Size per head is not a multiple of X (TensorRT-LLM/cpp /tensorrt_llm/kernels/unfusedAttentionKernels.cu:1710) this error,`

I printed out that the hidden size and head nums of the openllama 3B model are 3200 and 32 respectively, so the headsize （Size per head ）is 100, which is not a multiple of 8. The kernel does not support it. What should I do?

**Comment by PerkzZheng** [comment]
you should pad weights' head size to 104 or 128 (note that fmha doesn't support 104 head size), and use the new head size when building the engines. 

Non-multiple-of-8 head sizes are not supported mainly because of performance concerns.

**Comment by jpf888** [comment]
> you should pad weights' head size to 104 or 128 (note that fmha doesn't support 104 head size), and use the new head size when building the engines.
> 
> Non-multiple-of-8 head sizes are not supported mainly because of performance concerns.

thanks you !

Ask a few more questions,  I understand that before building, should first use a script to pad the weight, and then the quantization (int8 scale) also needs to be redone using the new head size.

Is correct?  and does the unfusedAttentionKernels need to be changed?

**Comment by PerkzZheng** [comment]
>  I understand that before building, should first use a script to pad the weight, and then the quantization (int8 scale) also needs to be redone using the new head size.

Yes, you can think of that we have a new model checkpoint with padded weights.

> does the unfusedAttentionKernels need to be changed?

It is not needed since the head size is multiple of 8 now. TRT-LLM should only know it is 104 (it doesn't matter whether it is padded or not).

**Comment by jpf888** [comment]
> > I understand that before building, should first use a script to pad the weight, and then the quantization (int8 scale) also needs to be redone using the new head size.
> 
> Yes, you can think of that we have a new model checkpoint with padded weights.
> 
> > does the unfusedAttentionKernels need to be changed?
> 
> It is not needed since the head size is multiple of 8 now. TRT-LLM should only know it is 104 (it doesn't matter whether it is padded or not).

thanks！ 

I think this kernel is transpose for kv cache, so this error will appear in int8 kvcache. 

So this problem does not exist for GPTQ?  I can temporarily bypass this problem if I do GPTQ？

**Comment by PerkzZheng** [comment]
> I think this kernel is transpose for kv cache, so this error will appear in int8 kvcache.
So this problem does not exist for GPTQ? I can temporarily bypass this problem if I do GPTQ？

The generation MHA kernels also have the limitation of multiple-of-8 head size，so padding with zeros is the best solution for now.

**Comment by jpf888** [comment]
> > I think this kernel is transpose for kv cache, so this error will appear in int8 kvcache.
> > So this problem does not exist for GPTQ? I can temporarily bypass this problem if I do GPTQ？
> 
> The generation MHA kernels also have the limitation of multiple-of-8 head size，so padding with zeros is the best solution for now.

Hello, I have completed padding and can run it, but the sample result is wrong. I would like to ask, if my weight size is (32000, 3200), pad 3200 to 3584, size is (32000, 3584), so that each row A large number of 0s will be padded, but when the memory is read, the data is read row by row continuously. As a result, the data originally accessed in the second row now becomes 0. Therefore, the result is incorrect. Is this analysis correct?

**Comment by PerkzZheng** [comment]
You need to pad the head size, not the hidden size to multiple of 8. 
so, it should be like 100 + 4 zeros + 100 + 4 zeros + .....
At the same time, you need to modify the model config (like hidden_size or head_size to the padded one). TRT-LLM will take everything as head size 104, and have no ideas about the zero paddings.

**Comment by jpf888** [comment]
> You need to pad the head size, not the hidden size to multiple of 8. so, it should be like 100 + 4 zeros + 100 + 4 zeros + ..... At the same time, you need to modify the model config (like hidden_size or head_size to the padded one). TRT-LLM will take everything as head size 104, and have no ideas about the zero paddings.

sorry，I'm still a little confused
When building the engine in tensorllm, head size is not a separate weight, but only some attention, mlp, and embed_token weights. . So I can't directly pad the head size.

Before entering the kernel calculation, the head size is calculated through mHiddenSize and mNbHeads of a GptModelConfig class, so do i need to change it here?

I didn't find the location where pad head size can be found in llama build.py. Where is it?

**Comment by PerkzZheng** [comment]
by default, head size = hidden_size // num_attention_heads.

for qkv-projection weights, you need to pad the head size as I described.  for the dense weights (after self-attention), you need to also pad each head size in the K dimension of GEMM operation while only need to pad zeros to the end of hidden size in the N dimension.  by doing that padding strategy, we can make sure the results are correct.

Then, you can pad zeros to the end of hidden size for all other weights (forget about head size).

We are working on making this padding strategy part of the engine builder, but it needs extra weeks to have that feature as we have too many features in the list.  hopefully, we can release that soon.

**Comment by lix19937** [comment]
@PerkzZheng  In my opinion,   
https://github.com/NVIDIA/TensorRT-LLM/blob/v0.8.0/cpp/tensorrt_llm/kernels/unfusedAttentionKernels.cu#L1647      
```cpp 
void invokeTranspose4dBatchMajor(const T* kSrc, const T* vSrc, KVCacheBuffer& kvTable, const int localBatchSize,
    const int seqLen, const int attentionWindowSize, const int sizePerHead, const int localHeadNum,
    const KvCacheDataType cache_type, const float* kvScaleOrigQuant, const int* sequence_lengths, cudaStream_t stream)
{
    // Block handles both K and V tile.
    dim3 blockSz(128, 2);
    constexpr int x = (sizeof(T) == 4) ? 4 : 8;
    dim3 gridSz((seqLen * sizePerHead / x + blockSz.x - 1) / blockSz.x, localBatchSize, localHeadNum);

    TLLM_CHECK_WITH_INFO(sizePerHead % x == 0, "Size per head is not a multiple of X");
```
`sizePerHead`  from  https://github.com/NVIDIA/TensorRT-LLM/blob/v0.8.0/cpp/tensorrt_llm/plugins/gptAttentionCommon/gptAttentionCommon.cpp#L459  

`mHeadSize` from onnx node attribute, we can modify onnx node attribute to change `mHeadSize` to 104, and padding zeros for a series of weights accordingly.  

----------------------- 

Another way, we can close the check of https://github.com/NVIDIA/TensorRT-LLM/blob/v0.8.0/cpp/tensorrt_llm/kernels/unfusedAttentionKernels.cu#L1647    
```cpp     
TLLM_CHECK_WITH_INFO(sizePerHead % x == 0, "Size per head is not a multiple of X");
```  
and overload `transpose4dBatchMajorKVCache` for `sizePerHead % x != 0`  case.
